### PR TITLE
Add LDAP authentication support

### PR DESCRIPTION
**Pull request by dvdkon**
This is a minimal PR adding support for authentication via LDAP by trying to bind to an LDAP server if internal Mealie authentication fails. If the user is on the LDAP server but not in the Mealie database, a new user object is inserted into the database.

I've made no UI changes as they aren't strictly necessary, but they might be useful, like disabling password changes for LDAP users altogether or checking the current password also via LDAP.

### CODE REVIEW

File: mealie/core/config.py [hunk 1]
```diff
@@ -151,6 +151,11 @@ def public_db_url(cls, v: Optional[str], values: dict[str, Any]) -> str:
     DEFAULT_EMAIL: str = "changeme@email.com"
     DEFAULT_PASSWORD: str = "MyPassword"
 
+    USE_LDAP_AUTH: bool = False
```
Comment (hay-kot) [inline mealie/core/config.py:154]: Can you change `USE_LDAP_AUTH` to `LDAP_AUTH_ENABLED` to keep the prefix consistent with the other variables?

File: mealie/core/config.py [hunk 2]
```diff
@@ -151,6 +151,11 @@ def public_db_url(cls, v: Optional[str], values: dict[str, Any]) -> str:
...
+    LDAP_SERVER_URL: str = None
+    LDAP_BIND_TEMPLATE: str = None
+    LDAP_ADMIN_FILTER: str = None
```
Comment (hay-kot) [inline mealie/core/config.py:157]: Can you document these new variables in the `template.env` file and the list of env variables in the docs? 

https://github.com/hay-kot/mealie/blob/master/docs/docs/documentation/getting-started/install.md

File: pyproject.toml [hunk 3]
```diff
@@ -35,6 +35,7 @@ pathvalidate = "^2.4.1"
 apprise = "^0.9.2"
 recipe-scrapers = "^13.2.7"
 psycopg2-binary = {version = "^2.9.1", optional = true}
+python-ldap = {version = "^3.3.0", optional = true}
```
Comment (hay-kot) [inline pyproject.toml:38]: Why is this optional?

File: mealie/core/security.py [hunk 4]
```diff
@@ -11,6 +11,37 @@
...
+        user.admin = len(conn.search_s(
+            user_dn,
+            ldap.SCOPE_BASE,
+            settings.LDAP_ADMIN_FILTER,
+            [])) > 0
+        db.users.update(session, user.id, user)
+
+    return user
+
+
```
Comment (hay-kot) [inline mealie/core/security.py:44]: Can you add a brief doc comment on what this function does and why?

**Review by hay-kot — COMMENTED** [review]
Overall it looks good, I like that it's a module and can just plugin as necessary. It looks like you made the installation optional for `python-ldap`. This won't be included in our CI/CD, was that on purpose? 

Additionally, I'd like to see some sort of documentation and testing to ensure this works and keeps working further down the line. It does look like you can spin up an LDAP server for testing, but I'm not sure.

**Comment by andretheolauret** [comment]
Hello I use Mealie and want to integrate it in app catalog of the project YunoHost, but for better support they require to have LDAP support. Can this PR be more under development please

**Comment by dvdkon** [comment]
> Can this PR be more under development please

I *just* submitted it yesterday! It's good to know this functionality is wanted but please have patience, I'm sure we'll have it sorted out soon.

**Comment by andretheolauret** [comment]
> > Can this PR be more under development please
> 
> I _just_ submitted it yesterday! It's good to know this functionality is wanted but please have patience, I'm sure we'll have it sorted out soon.

Ahah sorry I haven’t seen the PR date, yeah I think adding to Yunohost project will permit lot of newly users but right now I need ldap and also http auth if possible

**Comment by dvdkon** [comment]
@hay-kot Thanks. I made the dependency optional, because LDAP auth is an optional part of the program that I don't think most users will use, and it was really easy just to conditionally import the package. I didn't know it means no CI.

I've added some documentation, the options are pretty standard so hopefully it'll be enough.

It would be nice to test this, but most LDAP servers are big beasts made with geo-distributed operation in mind. Probably the simplest option would be to create a quick node.js server with [ldapjs](http://ldapjs.org/). I can try my hand at it if adding another dependency isn't a problem. The tests I see are all Python, how would one go about adding a new language?

**Comment by hay-kot** [comment]
@dvdkon, ah got it. You'll have to include it as a standard dependency so it gets packaged with the build and is available for use. 

Can you use the libraries built in [slapdtest module](https://www.python-ldap.org/en/python-ldap-3.3.0/reference/slapdtest.html) to test it? Seems like that's what it's for, but I could be missing something.

**Comment by dvdkon** [comment]
All right, I'll change it to a full dependency.

I wasn't aware of `slapdtest`, it still needs the OpenLDAP server, but if it takes care of all the annoying setup for me, I'll use it to add a test.

**Comment by hay-kot** [comment]
> All right, I'll change it to a full dependency.
> 
> I wasn't aware of `slapdtest`, it still needs the OpenLDAP server, but if it takes care of all the annoying setup for me, I'll use it to add a test.

I had another thought. I don't think we need to test the LDAP connection/auth. We can mock the return from the function and just ensure that what we are doing in both cases is what we expect.

**Comment by dvdkon** [comment]
That sounds like a simpler option. Does Mealie use any mocking library right now?

**Comment by hay-kot** [comment]
> That sounds like a simpler option. Does Mealie use any mocking library right now?

No, I haven't had the need. Alternatively, you could also stick the LDAP library functions in a class and override the methods using the monkeypatch fixture in Pytest. Kind of clunky, but it would work.

**Comment by xionous** [comment]
Thank you @dvdkon for this PR, I eagerly await its integration so I can start using LDAP.

**Comment by phinnay** [comment]
+1 to say I'm also eagerly awaiting this feature, Love the software!!!

**Comment by dvdkon** [comment]
I've added a test using simple `monkeypatch` mocking and made `ldap` a non-optional dependency so it can run on CI.

**Comment by hay-kot** [comment]
@xionous and @phinnay comments like this don't follow the etiquette for Github. if you don't have anything to contribute to the conversation it's best to use the reaction to show support. I find [this article/link to be helpful](https://github.com/kossnocorp/etiquette)

**Comment by hay-kot** [comment]
@dvdkon LGTM. Looks like you need to run poetry update to get the CI/CD to run. I also just noticed that you're targeting the master branch. The workflow for this project is to contribute to the 'dev' branch. I should have caught it earlier to let you know. If you're able to change it and target the dev branch that would be extremely helpful, if not I can take a stab at it later this week. Planning to get a new release drafted by Sunday so if you can't get to it by Saturday I'll get it fixed up.

**Comment by JOduMonT** [comment]
@hay-kot just to understand, Mealie still don't work with LDAP right ?

**Comment by hay-kot** [comment]
@JOduMonT Correct, this PR was never merged. Looks like I dropped the ball on changing the target branch and the OP didn't either. I'll try to get back to it, but I'm not sure when that will be.

**Comment by dvdkon** [comment]
I hadn't realised this PR had been closed, I didn't have time to finish it and hoped @hay-kot would fix the CI issues and rebase the PR. I'll try to finish it myself soon, the code is running on my server already so I should take care of it :)

**Comment by jado** [comment]
Great App! LDAP support is great for SSO at home.
Unfortunately I was not able to the server via ldaps when I set

`LDAP_SERVER_URL: ldaps://ipa.example.com
`
I get:
`
ldap.SERVER_DOWN: {'result': -1, 'desc': "Can't contact LDAP server", 'ctrls': [], 'info': (unknown error code)'}
`
python-ldap supports LDAPS, any pointers if mealie supports it?

**Pull request closed without merging**